procedural/constraints/src/main/kotlin/gov/nasa/jpl/aerie/constraints/Violations.kt

```diff
@@ -0,0 +1,85 @@
+package gov.nasa.jpl.aerie.constraints
+
+import gov.nasa.jpl.aerie.constraints.ActivityId.DirectiveId
+import gov.nasa.jpl.aerie.constraints.ActivityId.InstanceId
+import gov.nasa.jpl.aerie.constraints.Violations.Companion.getActivityId
+import gov.nasa.jpl.aerie.timeline.BaseTimeline
+import gov.nasa.jpl.aerie.timeline.BoundsTransformer
+import gov.nasa.jpl.aerie.timeline.Timeline
+import gov.nasa.jpl.aerie.timeline.collections.Instances
+import gov.nasa.jpl.aerie.timeline.collections.Intervals
+import gov.nasa.jpl.aerie.timeline.collections.Windows
+import gov.nasa.jpl.aerie.timeline.collections.profiles.Real
+import gov.nasa.jpl.aerie.timeline.ops.*
+import gov.nasa.jpl.aerie.timeline.ops.coalesce.CoalesceNoOp
+import gov.nasa.jpl.aerie.timeline.payloads.IntervalLike
+import gov.nasa.jpl.aerie.timeline.payloads.activities.Activity
+import gov.nasa.jpl.aerie.timeline.payloads.activities.Directive
+import gov.nasa.jpl.aerie.timeline.payloads.activities.Instance
+import gov.nasa.jpl.aerie.timeline.util.preprocessList
+
+/** A timeline of [Violations][Violation]. */
+data class Violations(private val timeline: Timeline<Violation, Violations>):
+    Timeline<Violation, Violations> by timeline,
+    ParallelOps<Violation, Violations>,
+    NonZeroDurationOps<Violation, Violations>,
+    CoalesceNoOp<Violation, Violations>
+{
+  constructor(vararg violation: Violation): this(violation.asList())
+  constructor(violations: List<Violation>): this(BaseTimeline(::Violations, preprocessList(violations, null)))
+
+  /**
+   * Maps the list of associated activity ids on each violation.
+   *
+   * @param f a function which takes a [Violation] and returns a new list of ids.
```

Comment:
Continuation of my previous question, what would an example `f: Violation -> List<ActivityId>` be?

Reply:
Keep it!